llvm/lib/CodeGen/GlobalISel/CallLowering.cpp

```diff
@@ -790,6 +805,62 @@ bool CallLowering::handleAssignments(ValueHandler &Handler,
       CCValAssign &VA = ArgLocs[j + Idx];
       const ISD::ArgFlagsTy Flags = Args[i].Flags[Part];
 
+      // We found an indirect parameter passing and we are at the first part of
+      // the value being passed. In this case copy the incoming pointer into a
+      // virtual register so later we can load it.
+      if (VA.getLocInfo() == CCValAssign::Indirect && Flags.isSplit()) {
+        IndirectParameterPassingHandled = true;
+        bool IsInStack = false;
+        Register PhysReg;
+        if (VA.isRegLoc()) {
+          PhysReg = VA.getLocReg();
+        } else if (VA.isMemLoc()) {
+          IsInStack = true;
+          LLT MemTy = Handler.getStackValueStoreType(DL, VA, Flags);
+          MachinePointerInfo MPO;
+          PhysReg = Handler.getStackAddress(
```

Comment:
This isn't right. getStackAddress returns a virtual register, and that shouldn't be passed to assignValueToReg. I guess this is how you get a virtual register in the live-ins list.
assignValueToReg should only be called if VA.isRegLoc() returns true.

I think that passing a pointer containing the address of an indirect object can be handled the same way as passing an ordinary pointer argument (one without the `Indirect` flag set). The `Indirect` flags only tells us that we should do some _additional_ work. I'd try to remove the `break` below and let the generic code handle the pointer.

Reply:
I'd also suggest to look into how `sret` arguments are handled, as they are also passed indirectly.

Reply:
The virtual live-in register problem is fixed.
I have tried to do the restructure the code to use the generic parts, instead of re-implementing them.

I had one case, where I could not do, that. That is the case when the pointer to stack is passed in a register, since there is no other branch in the later coming code that would assign the pointer to stack to the register used in parameter passing.